{% extends 'NetExplorer/base.html' %}
{% load static %}

{% block title %}- Tutorial{% endblock title %}



{% block content %}

<div class="row">
    <nav class="col-sm-2" id="myScrollspy">
        <ul class="nav nav-pills nav-stacked" data-spy="affix" data-offset-top="420">
            <li><a href="#genesearch">Gene Search</a></li>
            <li><a href="#netexplorer">NetExplorer</a></li>
            <li><a href="#pathwayfinder">Pathway Finder</a></li>
            <li><a href="#planexp">PlanExp</a></li>
            <hr>
            <li id="back_home"><a href="{% url 'index' %}" ><span class="glyphicon glyphicon-triangle-left"></span>Go back Home</a></li>
        </ul>
    </nav>

    <div class="col-sm-10">

      <!-- GENE SEARCH TUTORIAL -->
      <div class="tutorial-panel panel panel-info">
        <div class="panel-heading"><a class="tutorial-title" name="genesearch"><h3>Gene Search</h3></a></div>
        <div class="panel-body">
            <div class="panel panel-default">
                <div class="panel-heading">
                    Table of Contents
                </div>
                <div class="panel-body">
                    <ol>
                        <li class="tutorial-li"><a href="#genesearch-search">Search form</a></li>
                        <li class="tutorial-li"><a href="#genesearch-results">Results</a></li>
                        <li class="tutorial-li"><a href="#genesearch-card">Information card</a></li>
                    </ol>
                </div>
            </div>

            <p>
                <strong>GeneSearch</strong> allows users to look for proteins/contigs across the different planarian transcriptomes and Human reference interactome.
            </p>
            <br>
            <h4 id="genesearch-search">Search form</h4>
            <img class="tutorial-subtitle-netexplorer" src="{% static 'Images/tutorial/underline-genesearch.png' %}"/>
            <br>
            <img class="tutorial-img" src="{% static 'Images/tutorial/genesearch-form.png' %}"/>
            <br>
            <p>
                In order to search contigs/proteins, you have to select one of the available datasets (planarian or Human). Then, write one of the following types of symbols:
            </p>

            <ul>
                <li class="tutorial-li">
                    <strong>Contig identifiers</strong> (which will be specific to each transcriptome). E.g. <i>cth1_Trc_comp11081_c0_seq1</i> (for dataset <i>Cthulhu</i>)
                </li>

                <li class="tutorial-li">
                    <strong>Human protein symbols</strong>. This will retrieve all the homologous contigs to the specified human protein. E.g. <i>THOC2</i>.
                    You can use the wildcard <i>*</i>. E.g. <i>THOC*</i> will display the nodes homologous to <i>THOC, THOC1, THOC2, THOC3, THOC4, THOCXX, etc.</i>
                </li>

                <li class="tutorial-li">
                    <strong>Gene Ontology identifiers</strong>. This will display all the planarian contigs with a human homolog with the specified GO code.
                    E.g. <i>GO:0000002</i>.
                </li>

                <li class="tutorial-li">
                    <strong>PFAM identifiers</strong>. This will retrieve all the planarian contigs with the specified PFAM annotated through our protocol (that is, with HMMER with a e-value cutoff of 10<sup>-10</sup>).
                    E.g. <i>PF01369.15</i>.
                </li>
            </ul>

            <p>
                You can do multiple searches at once, using commas ',' to separate each item. You can mix the different types of symbols in the same search, e.g. <i>PF01369.15,GO:0000002</i>.
            </p>

            <br>
            <h4 id="genesearch-results">Results</h4>
            <img class="tutorial-subtitle-netexplorer" src="{% static 'Images/tutorial/underline-genesearch.png' %}"/>
            <br>
            <img class="tutorial-img" src="{% static 'Images/tutorial/genesearch-results.png' %}"/>
            <p>
                If there are proteins/contigs in the selected dataset that satisfy the search criteria, you will get a list of symbols inside blue boxes (see image). The symbol on the top right of each box always represents the database to which the symbol belongs. If the selected dataset was a Planarian transcriptome, the Human homolog will be specified inside the box, as can be seen in the image.
            </p>

            <p>
                Click on any of the boxes to get an <strong>information card</strong>, which will contain a summary of the information stored for that particular protein or contig.
            </p>

            <br>
            <h4 id="genesearch-card">Information card</h4>
            <img class="tutorial-subtitle-netexplorer" src="{% static 'Images/tutorial/underline-genesearch.png' %}"/>
            <br>
            <img class="tutorial-img" src="{% static 'Images/tutorial/genesearch-card1.png' %}"/>
            <br>
            <p>
                An information card is a brief summary of a contig or protein stored in any of our interactomes. Here, you will be able to download FASTA sequences, look for the annotated homolog for each contig, look for annotated PFAM domains, GO codes, and predicted interactions.
            </p>
            <p>
                You can navigate through different contigs/proteins without losing the results from GeneSearch, NetExplorer or PathwayFinder: only the information card will change. You can also navigate back using your browser's back button <span class="tutorial-btn">‚Üê</span>.
            </p>
            <p>
                Close the card using your <span class="tutorial-btn">Esc</span> key or the <span class="tutorial-btn tutorial-highlight-genesearch">x</span> button on the top right. If you want to see the card in full-screen mode in a new tab (for example, to print it), click on <span class="tutorial-btn tutorial-highlight-genesearch"><span class="glyphicon glyphicon-new-window"></span></span>.
            </p>


        </div>
      </div>

       <!-- NETEXPLORER TUTORIAL -->
    <div class="tutorial-panel panel panel-success">
        <div class="panel-heading"><a class="tutorial-title" name="netexplorer"><h3>Net Explorer</h3></a></div>
        <div class="panel-body">
            <div class="panel panel-default">
                <div class="panel-heading">
                    Table of Contents
                </div>
                <div class="panel-body">
                    <ol>
                        <li class="tutorial-li"><a href="#netexplorer-search">Search Functionality</a></li>
                        <li class="tutorial-li"><a href="#netexplorer-basiccontrols">Basic Controls</a></li>
                            <ol>
                                <li class="tutorial-li"><a href="#netexplorer-basiccontrols-layout">Layout</a></li>
                                <li class="tutorial-li"><a href="#netexplorer-basiccontrols-homologs">Show homologs</a></li>
                                <li class="tutorial-li"><a href="#netexplorer-basiccontrols-connections">Connections manipulation</a></li>
                                <li class="tutorial-li"><a href="#netexplorer-basiccontrols-filter">Filter interactions</a></li>
                            </ol>
                        <li class="tutorial-li"><a href="#netexplorer-click">Click Behaviour</a></li>
                        <li class="tutorial-li"><a href="#netexplorer-expression">Expression Data</a></li>
                        <li class="tutorial-li"><a href="#netexplorer-controls">General Controls</a></li>
                    </ol>
                </div>
            </div>

            <img class="tutorial-animation" src="{% static 'Images/tutorial/netexplorer-animation-1.gif' %}">
            <p>
                NetExplorer allows users to explore the predicted networks dinamically, adding and removing nodes from several interactomes.
                Use the menus on the right to manipulate the graph, which will display on the left side of the screen. Each node on the graph
                will represent a contig (for planarian datasets) or a protein (for the Human reference dataset). These nodes will be connected by edges (lines),
                and each type of them will have a different meaning:
            </p>
            <img class="tutorial-img" src="{% static 'Images/tutorial/legend.png' %}"/>

            <br>
            <h4 id="netexplorer-search">Search functionality</h4>
            <img class="tutorial-subtitle-netexplorer" src="{% static 'Images/tutorial/underline-netexplorer.png' %}"/>
            <br>
            <img class="tutorial-img" src="{% static 'Images/tutorial/search.png' %}"/>
            <p>
                In order to search for contigs, you have to select a Dataset (a planarian transcriptome for which we predicted an interactome), and then look for contigs.
                You can use 4 types of symbols to do that:
            </p>
            <ul>
                <li class="tutorial-li">
                    <strong>Contig identifiers</strong> (which will be specific to each transcriptome). E.g. <i>cth1_Trc_comp11081_c0_seq1</i> (for dataset <i>Cthulhu</i>)
                </li>

                <li class="tutorial-li">
                    <strong>Human protein symbols</strong>. This will retrieve all the homologous contigs to the specified human protein. E.g. <i>THOC2</i>.
                    You can use the wildcard <i>*</i>. E.g. <i>THOC*</i> will display the nodes homologous to <i>THOC, THOC1, THOC2, THOC3, THOC4, THOCXX, etc.</i>
                </li>

                <li class="tutorial-li">
                    <strong>Gene Ontology identifiers</strong>. This will display all the planarian contigs with a human homolog with the specified GO code.
                    E.g. <i>GO:0000002</i>.
                </li>

                <li class="tutorial-li">
                    <strong>PFAM identifiers</strong>. This will retrieve all the planarian contigs with the specified PFAM annotated through our protocol (that is, with HMMER with a e-value cutoff of 10<sup>-10</sup>).
                    E.g. <i>PF01369.15</i>.
                </li>
            </ul>

            <p>
                You can do multiple searches at once, using commas ',' to separate each item. You can mix the different types of symbols in the same search, e.g. <i>PF01369.15,GO:0000002</i>.
            </p>
            <p>
                Once you have typed the symbols you want to look for, and have selected the appropiate dataset, click on the button <span class="tutorial-btn tutorial-highlight-netexplorer">+ Add</span>.
                Note that you will not lose the network already displayed on the left, so you can add nodes individually without losing the visualization.
            </p>
            <p>
                All the nodes for which you have searched (either using this search function or by "expanding" them, see below) will be outlined in green,
                so that you can remember them.
            </p>

            <br>
            <h4 id="netexplorer-basiccontrols">Basic controls</h4>
            <img class="tutorial-subtitle-netexplorer" src="{% static 'Images/tutorial/underline-netexplorer.png' %}"/>
            <br>

            <h4 id="netexplorer-basiccontrols-layout">1. Layout</h4>
            <img class="tutorial-img" src="{% static 'Images/tutorial/layout.png' %}"/>
            <p>Using this dropdown menu you can select the layout of the network. When you select it, the layout will be run on the display,
                and the positions of the nodes will change. Take note that some layouts will change each time you run them. There are 6 available layouts right now:</p>
            <img class="tutorial-img" src="{% static 'Images/tutorial/layouts.png' %}"/>
            <br>
            <hr>
            <h4 id="netexplorer-basiccontrols-homologs">2. Show homologs</h4>
            <img class="tutorial-img" src="{% static 'Images/tutorial/homologs.png' %}"/>
            <p>
                When changing this switch to <span class="tutorial-btn tutorial-highlight-netexplorer">On</span>, the homologs of the displayed planarian contigs will be drawn on the visualization.
                They will appear as blue diamonds. They behave as all the other nodes, but with the exception that you can't <span class="tutorial-btn tutorial-highlight-netexplorer">Expand</span> them.
                Turning this option <span class="tutorial-btn tutorial-highlight-netexplorer">On</span> or <span class="tutorial-btn tutorial-highlight-netexplorer">Off</span> will <strong>not</strong> change the layout of the planarian nodes.
            </p>
            <img class="tutorial-img" src="{% static 'Images/tutorial/homolog.png' %}"/>
            <br>
            <hr>
            <h4 id="netexplorer-basiccontrols-connections">3. Connections manipulation</h4>
            <img class="tutorial-img" src="{% static 'Images/tutorial/connections.png' %}"/>
                 <br>
                <span class="tutorial-btn"  class="tutorial-highlight-button">Show connections</span>: <br>
                    <p>
                        This button will add all the interactions between the nodes displayed on the visualization. It will NOT add any node.
                    </p>
                <br>
                <span class="tutorial-btn"  class="tutorial-highlight-button">Remove Unconnected</span>: <br>
                    <p>
                        Activating this button will remove all the nodes without a <strong>visible</strong> connection.
                        Note that this applies to all types of connection (homology or interactions).
                        Use together with <span class="tutorial-btn"  class="tutorial-highlight-button">Filter interactions</span> to filter networks.
                    </p>
            <br>

            <hr>
            <h4 id="netexplorer-basiccontrols-filter">4. Filter interactions</h4>
            <img class="tutorial-img" src="{% static 'Images/tutorial/filter.png' %}"/>
            <p>
                With this slider you can hide the interactions using the percentage of votes of the random forest classifier as a cut-off.
                Note that the hidden interactions are not removed, so you can get them back by using another cut-off.
            </p>
            <p>
                Right now, the only available cut-offs are: <i>0.6, 0.65, 0.70, 0.75, 0.8, 0.85, 0.9</i> and <i>0.95</i>.
                You can use this function in conjuction with the <span class="tutorial-btn"  class="tutorial-highlight-button">Remove Unconnected</span> button to filter the graphs; first filtering the
                interactions by the desired cut-off, and then removing all the unconnected nodes.
                This is useful when wanting to work with a more accurate network, or when the nodes displayed on the visualization is too high.
            </p>
            <br>


            <h4 id="netexplorer-click">Click behaviour</h4>
            <img class="tutorial-subtitle-netexplorer" src="{% static 'Images/tutorial/underline-netexplorer.png' %}"/>
            <br>

            <img class="tutorial-img" src="{% static 'Images/tutorial/behaviour.png' %}"/>
            <p>
                If you have nodes on the visualization, you can determine what the application will do when you click on them.
                There are 3 possible behaviours, and you can change them by clicking on the appropiate one:
            </p>
            <hr>
            <h4>1. Card</h4>
            <p>
                When this behaviour is set, an information card for the clicked contig/protein will appear on the screen.
                These cards are the same displayed in the <a href="#genesearch">Gene Search</a> application.
            </p>
            <hr>
            <br>
            <h4>2. Expand</h4>
            <img class="tutorial-img" src="{% static 'Images/tutorial/expand.png' %}"/>
            <p>
                When clicking on nodes you will add all the interactions for such node to the visualization.
            </p>
            <hr>
            <br>
            <h4>3. Delete</h4>
            <p>
                Delete the clicked node and all its interactions. Note that all the other nodes displayed on the graph
                will NOT be removed, even if they have no other interactions.
            </p>

            <hr>
            <br>
            <h4 id="netexplorer-expression">Expression Data</h4>
            <img class="tutorial-subtitle-netexplorer" src="{% static 'Images/tutorial/underline-netexplorer.png' %}"/>
            <br>
            <img class="tutorial-img" src="{% static 'Images/tutorial/mapexpression.png' %}"/>
            <p>
                If you click on the <span class="tutorial-btn"  class="tutorial-highlight-button">Map Expression</span> button, the dialog showed on the above image will appear.
                Using this dialgo, you will be able to select a Expression experiment and map the expression values over the displayed nodes on the visualization.
                There are two ways in which you can do this: mapping the values of one sample as a color gradient, or mapping the logFC of two samples as a divergent color gradient.
                Make sure you are selecting two different samples in the <span class="tutorial-btn"  class="tutorial-highlight-button">Two Samples</span> mode.
            </p>
            <img class="tutorial-img" src="{% static 'Images/tutorial/samples.png' %}"/>
            <p>
                Finally, select a color gradient from the available presets and click <span class="tutorial-btn tutorial-highlight-netexplorer">OK</span>.
                The process of fetching the expression values and coloring the nodes may take a while depending on the number of nodes displayed.
                None of this will change the layout of the nodes, so feel free to experiment with the colors and the samples.
                When the process is done, a legend will appear on the visualization:
            </p>
            <img class="tutorial-img" src="{% static 'Images/tutorial/expressionlegend.png' %}"/>
            <p>
                If you are in <span class="tutorial-btn"  class="tutorial-highlight-button">One Sample</span> mode, each color will represent the percentil
                to which the expression value belongs; for example, the color for the percentil 10% indicates that the nodes colored like that
                have a expression value lower than 95% to 90% of the genes in the gene expression experiment.
            </p>

            <p>
                If you are in <span class="tutorial-btn"  class="tutorial-highlight-button">Two Samples</span> mode, each color will represent the log2FoldChange between
                the selected <i>sample 1</i> and <i>sample 2</i>.
            </p>

            <p>
                In either case, the nodes colored as NA are contigs for which there is no expression information for that particular experiment and sample.
            </p>
            <p>
                When a expression experiment has been mapped onto the visualization you can keep adding nodes normally. However, if you want the expression
                to be mapped to the new added nodes, you have to open the dialog again and re-run the process. Don't expect new nodes to be colored according to
                the expression information previously mapped.
            </p>
            <br>


            <h4 id="netexplorer-controls">General Controls</h4>
            <img class="tutorial-subtitle-netexplorer" src="{% static 'Images/tutorial/underline-netexplorer.png' %}"/>
            <br>
            <img class="tutorial-img" src="{% static 'Images/tutorial/controls.png' %}"/>
            <p>
                These simple buttons allow users to save, import and export the visualization in several formats, as well as centering it to the screen.
            </p>

            <div class="tutorial-btn"  class="tutorial-highlight-button">
                <span class="glyphicon glyphicon-resize-small" aria-hidden="true"></span>
            </div>&nbsp;&nbsp; Center the visualization when clicked.
            <br><br>

            <div class="tutorial-btn"  class="tutorial-highlight-button">
                <span class="glyphicon glyphicon-picture" aria-hidden="true"></span>
            </div>&nbsp;&nbsp; Saves a png image of the visualization. Please, take note that the expression legend will NOT be saved, as it is not part of the visualization.
            <br><br>

            <div class="tutorial-btn"  class="tutorial-highlight-button">
                <span class="glyphicon glyphicon-share" aria-hidden="true"></span>
            </div>&nbsp;&nbsp; Exports the graph (nodes and edges) as a tabular file. This graph can then be imported to Cytoscape, or Gephi, or other programs to perform
            further analyses.
            <br><br>

            <div class="tutorial-btn"  class="tutorial-highlight-button">
                <span class="glyphicon glyphicon-save" aria-hidden="true"></span>
            </div>&nbsp;&nbsp; Saves the graph as a JSON file. This is the format used by NetExplorer. If you want to save a particular visualization to work on it later in
            NetExplorer, use this button. All the positions, colors, nodes and edges will be saved and they can be restored.
            <br><br>

            <div class="tutorial-btn"  class="tutorial-highlight-button">
                <span class="glyphicon glyphicon-open" aria-hidden="true"></span>
            </div>&nbsp;&nbsp; Imports a JSON file (exported by using the previous button). All the visualization will be restored. Any network on the visualization previous
            to importing the JSON file will be lost, so please be careful.
            <br><br>

            <div class="tutorial-btn"  class="tutorial-highlight-button">
                <span class="glyphicon glyphicon-trash" aria-hidden="true"></span>
            </div>&nbsp;&nbsp; Remove everything (nodes, edges, colors, legends) from the visualization. This process is <strong>irreversible</strong>.
            <br><br>



        </div>
    </div>

       <!-- PATHWAY FINDER TUTORIAL -->
    <div class="tutorial-panel panel panel-warning">
        <div class="panel-heading"><a class="tutorial-title"  name="pathwayfinder"><h3>Pathway Finder</h3></a></div>
        <div class="panel-body">
            <div class="panel-body">
                <div class="panel panel-default">
                    <div class="panel-heading">
                        Table of Contents
                    </div>
                    <div class="panel-body">
                        <ol>
                            <li class="tutorial-li"><a href="#pathwayfinder-form">Search Form</a></li>
                            <li class="tutorial-li"><a href="#pathwayfinder-results">Results</a></li>
                        </ol>
                    </div>
                </div>
            <img class="tutorial-animation" src="{% static 'Images/tutorial/pathfinder-animation-1.gif' %}">
            <p>
                PathwayFinder lets you select two (or more) nodes and retrieve all the possible paths of a given length between them in
                a particular interactome. Right now, due to performance issues on the client side, the user can only select up to 5 edges
                separating the given nodes.
            </p>
            <br>
            <h4 id="pathwayfinder-form">Search Form</h4>
            <img class="tutorial-subtitle-pathfinder" src="{% static 'Images/tutorial/underline-pathfinder.png' %}"/>
            <br>
            <img class="tutorial-img tutorial-float" src="{% static 'Images/tutorial/pfind-form.png' %}"/>
            <p>
                With this form you can select the <span class="tutorial-btn"  class="tutorial-highlight-button"><img class="pfind-form-icon tutorial-flag" src="{% static 'Images/starting-node-icon.png' %}"/>StartingNode</span> and the <span class="tutorial-btn"  class="tutorial-highlight-button"><img class="pfind-form-icon tutorial-flag" src="{% static 'Images/ending-node-icon.png' %}"/>FinalNode</span>.
            </p>
            <p>
                As in other searches in PlanNET, you can use:
                <ul>
                    <li class="tutorial-li">Contig identifiers.</li>
                    <li class="tutorial-li">Human protein symbols (including <i>*</i> wildcards)-.</li>
                    <li class="tutorial-li">PFAM accession codes.</li>
                    <li class="tutorial-li">GeneOntology codes.</li>
                </ul>
                Be aware that you can end up with several <span class="tutorial-btn"  class="tutorial-highlight-button"><img class="pfind-form-icon tutorial-flag" src="{% static 'Images/starting-node-icon.png' %}"/>StartingNodes</span> and <span class="tutorial-btn"  class="tutorial-highlight-button"><img class="pfind-form-icon tutorial-flag" src="{% static 'Images/ending-node-icon.png' %}"/>FinalNodes</span>.
                This will increase the potential number of results. Because each pathway will be a different visualization, be aware that the performance will depend on the computer power on the client side.
                For that reason, it is advisable to first select a small Path Length and try higher lengths depending on the results.
            </p>
            <p>
                Once everything is selected, press the button <span class="tutorial-btn tutorial-highlight-pathfinder">FIND</span>. The results will appear on the left.
            </p>
            <br><br><br>

            <br>
            <h4 id="pathwayfinder-results">Results</h4>
            <img class="tutorial-subtitle-pathfinder" src="{% static 'Images/tutorial/underline-pathfinder.png' %}"/>
            <br>
            <img class="tutorial-img" src="{% static 'Images/tutorial/pfind-results.png' %}"/><br>
            <p>
                On the top of the screen will appear a summary of the search and the number of results (in this case, 508 pathways).
            </p>
            <p>
                All the pathways will be sorted by the mean of the votes of all the interactions in it. That is, by the confidence.
            </p>
            <p>
                You can filter the pathways with specific contigs using the above search box and by clicking on <span class="tutorial-btn tutorial-highlight-pathfinder">Filter</span>.
                You can use multiple contig identifiers by using commas ','. You can <strong>only</strong> use planarian contig identifiers here.
            </p>
            <p>
                There are two buttons on top of each pathway:
            </p>
            <div class="tutorial-btn"  class="tutorial-highlight-button">
                <span class="glyphicon glyphicon-resize-small" aria-hidden="true"></span>
            </div>&nbsp;&nbsp; Center the path on the visualization.
            <br><br>
            <div class="tutorial-btn"  class="tutorial-highlight-button">
                <img src="{% static 'Images/netexplorer-mini.png' %}" class="netexplorer-mini"/>
            </div>&nbsp;&nbsp; Open a new tab with the pathway on NetExplorer.
            <br><br>
            <p>
                You can click on the nodes to display an information card.
            </p>
        </div>

    </div>
</div>

        <div class="tutorial-panel panel panel-danger">
            <div class="panel-heading"><a class="tutorial-title"  name="planexp"><h3>PlanExp</h3></a></div>
            <div class="panel-body">
                <div class="panel-body">
                    <div class="panel panel-default">
                        <div class="panel-heading">
                            Table of Contents
                        </div>
                        <div class="panel-body">
                            <ol>
                                <li class="tutorial-li"><a href="#planexp-form">Experiment selection</a></li>
                                <li class="tutorial-li"><a href="#plot-controls">Plot controls</a></li>
                                <li class="tutorial-li"><a href="#planexp-dge">Differential gene expression</a></li>
                                <li class="tutorial-li"><a href="#planexp-gplot">Gene expression plot</a></li>
                                <li class="tutorial-li"><a href="#planexp-coexpression">Gene co-expression</a></li>
                                <li class="tutorial-li"><a href="#planexp-tutorial-tsne">t-SNE plot</a></li>
                                <li class="tutorial-li"><a href="#planexp-tutorial-markers">Marker genes</a></li>
                                <li class="tutorial-li"><a href="#planexp-tutorial-regulatory-interactions">Genes Regulatory Interactions</a></li>
                                <li class="tutorial-li"><a href="#planexp-tutorial-network">Network</a></li>
                            </ol>
                        </div>
                    </div>
                    <img class="tutorial-animation" src="{% static 'Images/tutorial/planexp-animation-1.gif' %}">
  
                    <br>
                    <h4 id="planexp-form">Experiment Selection</h4>
                    <img class="tutorial-subtitle-planexp" src="{% static 'Images/tutorial/underline-planexp.png' %}"/>
                    <br>
                    <img class="tutorial-img" src="{% static 'Images/tutorial/planexp-expselection.png'%}" />
                    <p>
                        Use the first dropdown menu to select the RNA-seq experiment you want to explore. 
                        Each experiment will be mapped over different datasets (transcriptomes or genome references): 
                        use the second dropdown to select from the available ones for the selected experiment.
                    </p>

                    <img class="tutorial-img" src="{% static 'Images/tutorial/planexp-summary.png'%}" />
                    <p>
                        Once an experiment has been selected, a summary will appear on the application. 
                        The conditions on the right side correspond to the different factors of the experiment, which could be 
                        experimental factors (such as "Section" and "Time", as in the image), or technical ones (such as "Batch" or others).
                        Click on the conditions names to see the "levels" of that "factor".
                    </p>

                    <img class="tutorial-img" src="{% static 'Images/tutorial/planexp-toc.png'%}" />
                    <p>
                        While selecting the Experiment and the Dataset, different sections will appear in the table of contents section on the right.
                        Click on the desired section title to scroll automatically to the corresping section in the application.
                    </p>

                    <br>
                    <h4 id="plot-controls">Plot Controls</h4>
                    <img class="tutorial-subtitle-planexp" src="{% static 'Images/tutorial/underline-planexp.png' %}"/>
                    <br>
                    <img class="tutorial-img" src="{% static 'Images/tutorial/planexp-plot-controls.png'%}" />

                    <p>
                        All plots created by PlanExp are made with <a href="https://plot.ly/javascript/" target="_blank">Plotly.js</a>, 
                        and they all share the controls shown on the image. Click on each button to performed the corresponding action.
                    </p>

                    <ul id="tutorial-plot-controls-list">
                        <li><img src="{% static 'Images/tutorial/planexp-plot-img1.png'%}" /> Saves plot as a PNG image.</li>
                        <li><img src="{% static 'Images/tutorial/planexp-plot-img2.png'%}" /> Edit data in chart studio (external website).</li>
                        <li><img src="{% static 'Images/tutorial/planexp-plot-img3.png'%}" /> Zoom function (select area of graph to zoom-in).</li>
                        <li><img src="{% static 'Images/tutorial/planexp-plot-img4.png'%}" /> Pan function (move through the graph area).</li>
                        <li><img src="{% static 'Images/tutorial/planexp-plot-img5.png'%}" /> Box select.</li>
                        <li><img src="{% static 'Images/tutorial/planexp-plot-img6.png'%}" /> Lasso select.</li>
                        <li><img src="{% static 'Images/tutorial/planexp-plot-img7.png'%}" /> Zoom-in button.</li>
                        <li><img src="{% static 'Images/tutorial/planexp-plot-img8.png'%}" /> Zoom-out button.</li>
                        <li><img src="{% static 'Images/tutorial/planexp-plot-img9.png'%}" /> Autoscale.</li>
                        <li><img src="{% static 'Images/tutorial/planexp-plot-img10.png'%}" /> Default zoom level.</li>
                        <li><img src="{% static 'Images/tutorial/planexp-plot-img11.png'%}" /> Toggle Spike lines.</li>
                        <li><img src="{% static 'Images/tutorial/planexp-plot-img12.png'%}" /> Show closest data on hover.</li>
                        <li><img src="{% static 'Images/tutorial/planexp-plot-img13.png'%}" />Compare data on hover.</li>
                        
                    </ul>

                    <br>
                    <p>
                        When a trace legend appears on the right, either containing gene/transcript symbols or conditions, 
                        use the controls shown on the image below to controls which plot traces to show or hide.
                    </p>
                    <img class="tutorial-img" src="{% static 'Images/tutorial/planexp-plot-controls2.png'%}" />
                    
                    
                    <br>
                    <hr>
                    
                    <h4 id="planexp-dge">Differential Gene Expression</h4>
                    <img class="tutorial-subtitle-planexp" src="{% static 'Images/tutorial/underline-planexp.png' %}"/>
                    <br>
                    <img class="tutorial-img" src="{% static 'Images/tutorial/planexp-dge1.png'%}" />
                    <p>
                        The following section describes how to explore the differentially expressed genes across different conditions.
                        On the first dropdown "Condition" select the desired condition for which you want to compare levels. 
                        For example, in a time-course experiment consisting of one condition called "Time", select "Time" on the first dropdown.
                        Doing so will populate the two following dropdown menus with the appropiate levels.
                    </p>
                    <p>
                        In the case of the experiment "2013 Aboobaker Time-course", there are no comparisons between individual "Hour" levels, only the interaction 
                        between the conditions "Hour" and "Section" was computed. As such, select "Hour - Section" on the dropdown menu, and then select the levels you want 
                        to compare. 
                    </p>
                    <p>
                        When choosing one level on one of the two "Select conditions to compare" dropdowns, only the levels for which there are pair-wise comparisons 
                        available will be visible on the second dropdown menu. When both condition levels have been selected, a table and a volcano plot will appear below.
                    </p>
                    
                    <img class="tutorial-img" src="{% static 'Images/tutorial/planexp-dge4.png'%}" />
                    <br>
                     <img class="tutorial-img" src="{% static 'Images/tutorial/planexp-dge2.png'%}" />
                    <p>
                        This table will contain the differentially expressed genes between the two selected levels (in this case, "0h Head" vs "6h Head"). 
                        The adjusted p-value cut-offs will be the ones described in the respective research article of the experiment.
                    </p>
                    <p>
                        Keep in mind that the order of "Condition 1" and "Condition 2" will determine the meaning of the Log Fold Change direction.
                        The Log Fold Change will be computed as log ( Condition 1 / Condition 2), thus, 
                        positive fold changes will indicate over-expression in Condition 1 with respect to Condition 2, and negative fold changes under-expression.
                    </p>

                    <p>
                        Click on the names of transcripts, planarian genes or Human proteins to get a gene card with thorough information about its annotations, see <a href="#genesearch-card">information card</a>. 
                        Download the table by clicking on the <span class="tutorial-btn">CSV</span> button.
                    </p>

                    <img class="tutorial-img" src="{% static 'Images/tutorial/planexp-dge3.png'%}" />
                    
                    <p>
                        This volcano plot shows each differentially expressed gene in relation to its Fold change (x-axis) and its P-Value (y-axis). 
                        The higher the y-axis, the more significant the gene is in the comparison. 
                        Hover on the points of the plot to see the specific gene/transcript name and its values.
                        The plot controls are located on the top right of the visualization, see <a href="#plot-controls">plot controls</a>. 
                    </p>

                    <img class="tutorial-img" src="{% static 'Images/tutorial/planexp-goea.png'%}" />

                    <p>
                        Select a particular level from the previously selected comparison and click on the Run Analysis button 
                        to perform a Gene Ontology Enrichment Analysis. Download the results using the <span class="tutorial-btn">Download CSV</span> button, 
                        or download the images in each of the three available tabs (Biological Process, Molecular Function, and Cellular Component). 
                        The nodes in the networks, which represent the Gene Ontology Terms, are colored depending on their significance in the performed Fisher test.

                    </p>
                    
                    <br>
                    <hr>
                    
                    <h4 id="planexp-gplot">Gene Expression Plot</h4>
                    <img class="tutorial-subtitle-planexp" src="{% static 'Images/tutorial/underline-planexp.png' %}"/>
                    <br>

                    <p>
                        The Gene Expression Plot section will allow us to plot the gene expression of one, or several, genes 
                        across the different conditions of the selected experiment. For this tutorial, we will use the experiment "2018 Rajewsky Cell Atlas". 
                        Check how to interact with these plots in the <a href="#plot-controls">Plot Controls section</a>.
                    </p>

                    <img class="tutorial-img" src="{% static 'Images/tutorial/planexp-gplot1.png'%}" />

                    <p>
                        Write the genes you want to plot in the text input field. You can use planarian contig identifiers for the selected Dataset, 
                        planarian genes (SMESG...), human gene symbols from the <a href="https://www.genenames.org/" target="_blank">HGNC database</a>, PFAM domains (PF...), or Gene Ontology codes (GO:...). 
                    </p>
                    <p>
                        Select the Condition you want to visualize the expression values for, and each level of that condition 
                        will be positioned in the x-axis of the visualizations.
                    </p>
                    <br>

                    <h3>1. Heatmap</h4>
                    <img class="tutorial-img" src="{% static 'Images/tutorial/planexp-gplot2.png'%}" />

                    <p>
                        The Heatmap visualization will display the mean expression of each gene (row) in each condition (column) as a color gradient (right legend). 
                        Hover over the heatmap cells to see the expression values, and click on the transcript names to get a gene card and check their annotations.
                    </p>

                    <p>
                        <b>Keep in mind that, if plotting many genes or conditions at the same time, not all row or column names will be displayed, and you will have to use the Zoom-in function to explore the graph efectively, see <a href="#plot-controls">plot controls</a>. </b>
                    </p>
                    
                    <img class="tutorial-img" src="{% static 'Images/tutorial/planexp-gplot8.png'%}" />
                    <p>
                        In the case of Single-Cell experiments, the Condition option "Samples" will appear in the Condition dropdown menu. 
                        When selecting it and plotting it, a heatmap where each column corresponds to a single cell will be shown. This visualization 
                        is useful for identifying genes that are co-expressed in specific cells, and for comparing their expression levels. Contrary to the 
                        section <a href="#planexp-coexpression">Gene Co-expression</a>, this heatmap visualization allows for users to check more than two genes at the same time.
                    </p>

                    
                    <br>
                    <h3>2. Violin/Bar plot</h4>
                    <img class="tutorial-img" src="{% static 'Images/tutorial/planexp-gplot3.png'%}" />

                    <p>
                        The violin plot visualization allows users to see the distribution of expression values (y-axis) in each given condition (x-axis). 
                        It is equivalent to a flipped density plot. Because Single-Cell experiments tend to have many cells with zero expression for many genes, 
                        when exploring a single-cell experiment the option "Show only expressing cells" will appear. This option will remove cells that don't express 
                        the given genes, allowing the comparison of expression values in only those cells that are already expressing the genes.
                    </p>
                    <img class="tutorial-img" src="{% static 'Images/tutorial/planexp-gplot4.png'%}" />
                    <p>
                        When hovering on the violins, several statistics will be displayed: the 3rd quartile, the median and the minimum value.
                    </p>

                    <img class="tutorial-img" src="{% static 'Images/tutorial/planexp-gplot5.png'%}" />
                    <p>
                        In the case of experiments with only one sample per condition level (that is, without replicates) instead of a violin plot 
                        a bar plot will be displayed. The above bar plot was created by performing the same search in the "2015 Abril DGE" experiment 
                        on the Dataset "Adamidi".
                    <p>

                    <br>
                    <h3>3. Line Chart</h4>
                    <img class="tutorial-img" src="{% static 'Images/tutorial/planexp-gplot6.png'%}" />

                    <p>
                        When visualizing conditions that can be ordered (as is the case of the condition "Regenerating Day" of "2018 Rajewsky Cell Atlas"), 
                        it is useful to use the plot type "Line Chart".
                    </p>

                    <img class="tutorial-img" src="{% static 'Images/tutorial/planexp-gplot7.png'%}" />
                    <p>
                        When selecting an interaction Condition that includes an ordered condition (like time), and plotting a Line Chart, 
                        several line charts will be created for each level of the interaction. In the image above, where the condition 
                        "Day - Cluster" was selected, a line chart for each Cluster will be shown.  
                    </p>

                    <br>
                    <hr>
                    
                    <h4 id="planexp-coexpression">Gene Co-expression</h4>
                    <img class="tutorial-subtitle-planexp" src="{% static 'Images/tutorial/underline-planexp.png' %}"/>
                    <img class="tutorial-img" src="{% static 'Images/tutorial/planexp-coexpression.png'%}" />
                    <br>

                    <p>
                        The section Gene Co-expression will allow you to plot the expression of two genes in each sample of a Single-Cell Experiment 
                        and color it by a particular condition. Use the plot trace legend on the right to hide or show the desired conditions, see <a href="#plot-controls">plot controls</a>. 
                    </p>

                    <br>
                    <hr>
                    
                    <h4 id="planexp-tutorial-tsne">t-SNE plot</h4>
                    <img class="tutorial-subtitle-planexp" src="{% static 'Images/tutorial/underline-planexp.png' %}"/>
                    <img class="tutorial-img" src="{% static 'Images/tutorial/planexp-tsne1.png'%}" />
                    <br>

                    <p>
                        For single cell experiments, PlanExp allows users to create t-SNE plots. In them, cells are represented as points 
                        in a 2D space, where similar cells with a similar expression profile will tend to appear together. Keep in mind that 
                        the dimensionality reduction performed by t-SNE does not conserve the distance in the initial N-dimensional space. 
                        For further information, see <a href="https://en.wikipedia.org/wiki/T-distributed_stochastic_neighbor_embedding" target="_blank">this page</a>.
                    </p>

                    <p>
                        There are several options for plotting t-SNE visualizations for single cell experiments in PlanExp:
                    </p>

                    <img class="tutorial-img" src="{% static 'Images/tutorial/planexp-tsne2.png'%}" />

                    <ol>
                        <li>Coloring each cell by condition.</li>
                        <li>Coloring each cell by gene expression:</li>
                            <ul>
                                <li>Coloring cells by the expression of one gene.</li>
                                <li>Coloring cells by the mean expression of moltiple genes, and showing only those cells that express all of them.</li>
                            </ul>
                    </ol>

                    <p>
                        When plotting gene expression in the "Gene" tab, there will be two new options in a radio button (see image above). 
                        Users will be able to color cells (points) depending on their expression of one gene, or multiple genes. When selecting the 
                        second option "Colors by mean expression...", only those cells that express all the input genes will be shown. 
                        For instance, in the example below, only those cells expressing both ("dd_Smed_v6_740_0_1" and "dd_Smed_v6_332_0_1") will appear in the plot, 
                        and they are colored by the mean expression:
                    </p>
                    <pre>
                        (expression <sub>dd_Smed_v6_740_0_1</sub> + expression <sub>dd_Smed_v6_332_0_1</sub>) / 2
                    </pre>

                    <img class="tutorial-img" src="{% static 'Images/tutorial/planexp-tsne3.png'%}" />

                    <br>
                    <hr>
                    
                    <h4 id="planexp-tutorial-markers">Marker genes</h4>
                    <img class="tutorial-subtitle-planexp" src="{% static 'Images/tutorial/underline-planexp.png' %}"/>
                    <img class="tutorial-img" src="{% static 'Images/tutorial/planexp-markers.png'%}" />
                    <br>

                    <p>
                        Single cell experiments have a section, "Marker genes" in which genes that identify clusters of cells are listed as searchable tables.
                        These were computed by <a href="https://satijalab.org/seurat/" target="_blank">Seurat</a>'s AUC classifier. 
                        Select a cluster for which to retrieve markers, and then a table will appear. The AUC column is the Area Under the Curve in a 
                        <a href="https://en.wikipedia.org/wiki/Receiver_operating_characteristic" target="_blank">receiver operating characteristic</a> curve 
                        of the classifier built using the gene as a classifier of cells. A gene that, when expressed, identifies a cell as a member of a particular cluster with 100% accuracy 
                        has an AUC of 1. The Average difference column indicates if the classifier is built on the expression (e.g.: if a gene is more expressed than a particular value, 
                        classify this cell as a member of the cluster), or on its lack of expression (e.g.: if a genes is <i>less</i> expressed than a particular value, classify the cell as a member of the cluster).
                    </p>

                    <p>
                        Each table can be downloaded by clicking on the <span class="tutorial-btn">CSV</span> button.
                    </p>

                    <br>
                    <hr>
                    <h4 id="planexp-tutorial-regulatory-interactions">Gene Regulatory Interactions</h4>
                    <img class="tutorial-subtitle-planexp" src="{% static 'Images/tutorial/underline-planexp.png' %}"/>
                    <img class="tutorial-img" src="{% static 'Images/tutorial/planexp-regulatory-interactions.png'%}" />
                    <br>

                    <p>
                        We performed a prediction of gene regulatory interactions for single cell experiments. Regulatory interactions define 
                        relationships of activation or inhibition between one gene (regulator) which activates or inhibits the expression of another gene (target). 
                        These predictions were performed with the program <a href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0012776" target="_blank">GENIE3</a>.
                    </p>

                    <p>
                        <a href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0012776" target="_blank">GENIE3</a> requires 
                        users to define which genes are transcription factors (and thus, putative regulators of other genes). The selection of 
                        planarian genes as regulators was performed by looking at their PFAM annotations or at their GO annotations. The Source column 
                        of the generated table indicates if the regulator was decided to be a transcription factor by its PFAM domains or by its GO terms.
                    </p>

                    <p>
                        The score is a measure of how confident <a href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0012776" target="_blank">GENIE3</a> 
                        is of the prediction. However, this score does not have any statistical meaning.
                    </p>

                    <p>
                        Because the large number of regulatory links predicted by <a href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0012776" target="_blank">GENIE3</a>, 
                        we binned them in 10 categories, sorted by score. Users can select which group of regulatory interactions they want to visualize on the 
                        table by using the dropdown menu on the top left.
                    </p>

                    <p>
                        The button <span class="tutorial-btn">CSV</span> allows users to download the generated table, 
                        while the <span class="tutorial-btn">Send to Network viewer</span> sends the whole table of regulatory links to the network viewer (see <a href="#planexp-tutorial-network">next section</a>).
                    </p>


                    <br>
                    <hr>
                    
                    <h4 id="planexp-tutorial-network">Network</h4>
                    <img class="tutorial-subtitle-planexp" src="{% static 'Images/tutorial/underline-planexp.png' %}"/>
                    <img class="tutorial-img" src="{% static 'Images/tutorial/planexp-network1.png'%}" />
                    <br>

                    <p>
                        PlanExp network viewer allows users to map expression data onto regulatory networks. These can be 
                        either downloaded from PlanNET, retrieved from the <a href="#planexp-tutorial-regulatory-interactions">Regulatory Interactions</a> section, or 
                        created from scratch by users by using PlanExp <a href="#planexp-tutorial-network-editor">Network Editor</a>. 
                    </p>
                    
                    <br>
                    <h3 id="planexp-tutorial-network-controls">1. Network controls</h3>
                    <img class="tutorial-img" src="{% static 'Images/tutorial/planexp-network2.png'%}" />
                    <p>
                        The buttons above the Network viewer allow users to control the graph visualization in several ways. 
                    </p>

                    <ol>
                        <li><span class="tutorial-btn">Center</span>: Centers the visualization on the genes in the network. </li>
                        <li><span class="tutorial-btn">Layout Dropdown</span>: Select a different graph layout.</li>
                        <li><span class="tutorial-btn">Edit</span>: Opens the <a href="#planexp-tutorial-network-editor">Network Editor</a>.</li>
                        <li><span class="tutorial-btn">Import</span>: Imports JSON file with network (downloaded from PlanNET <a href="{% url 'net_explorer' %}" target="_blank">NetExplorer</a>, or from next button).</li>
                        <li><span class="tutorial-btn">Export</span>: Exports graph to JSON for later use.</li>
                        <li><span class="tutorial-btn">Save</span>: Saves a PNG image of the network.</li>
                        <li><span class="tutorial-btn">Delete</span>: Deletes the whole network.</li>
                    </ol>

                    
                    <br>
                    <h3 id="planexp-tutorial-network-editor">2. Network Editor</h3>
                    <img class="tutorial-img" src="{% static 'Images/tutorial/planexp-network-editor1.png'%}" />

                    <p>
                        The Network Editor allows for the creation of nodes (genes) and edges (interactions), as well as their deletion.

                        To add genes write a gene name on the top left input form and click on <span class="tutorial-btn">Add +</span>. 
                        Manually created nodes have a pink border color so that they can be easily spotted.
                    </p>
                    <p>
                        To add interactions hover on one of the genes you want to create an interaction for, and a small red dot will appear (see image above). 
                        Click on the red dot and drag to the other gene of the interaction. When you release the mouse button, an interaction will appear (violet edge in the image above).
                    </p> 
                    <p>
                        For deleting genes or interactions, click on the corresponding <span class="tutorial-btn">ON</span> switch button 
                        and then click on the gene or interaction that you want to remove.
                    </p>

                    <p>
                        Once you are happy with the network, click on <span class="tutorial-btn">Save</span> and the network will 
                        be sent to the main Network viewer.
                    </p>


                    <br>
                    <h3 id="planexp-tutorial-network-color">3. Color genes on network</h3>
                    <img class="tutorial-img" src="{% static 'Images/tutorial/planexp-network3.png'%}" />

                    <p>
                        On the right of the Network Viewer main section, we have the controls to color genes according to their expression.
                        Genes can be colored depending on their expression in <span class="tutorial-btn">One Condition</span> or <span class="tutorial-btn">Two Conditions:</span> (see a.1 and a.2). 
                    </p>

                    <p>
                        In the <span class="tutorial-btn">One Condition</span> tab, select the condition in the dropdown menu, the color profile, and the reference. 
                        The Network reference option will color the most expressed gene in the network as the strongest color in the color profile, and all the others 
                        according to their relation to that gene. The Experiment reference will color genes according to the gene with maximum expression in the experiment, 
                        even if that gene is not in the network. Genes will be colored in a gradient that goes from white (if they are not expressed) to strong red, blue or green (if they are highly expressed).
                    </p>

                    <p>
                        In the <span class="tutorial-btn">Two Conditions</span> tab, users will have to select two Conditions to compare them. 
                        The color gradients will be divergent instead of linear. Genes will be colored by their log Fold change using the selected gradient.
                    </p>

                    <p>
                        Independently of the option used, once the genes have been colored a color legend will appear on the left (b and c in image above). 
                        Hovering on the legend will highlight the genes that are expressed in a particular range (c in image above).
                    </p>

                    <br>
                    <h3 id="planexp-tutorial-network-color">4. Filter genes</h3>
                    <img class="tutorial-img" src="{% static 'Images/tutorial/planexp-network4.png'%}" />

                    <p>
                        Genes in the network can be filtered out (make them hidden) by selecting one or several conditions on the 
                        <span class="tutorial-btn">Select Conditions</span> dropdown menu (bottom right). Once the desired conditions 
                        have been selected, one can choose to filter the network by clicking on the <span class="tutorial-btn">Filter Network</span> button. 
                    </p>

                    <p>
                        The option <span class="tutorial-btn">Inclusive</span> will show all genes that are expressed in <b>at least 
                        one</b> of the selected conditions. The option <span class="tutorial-btn">Exclusive</span> will show only 
                        those genes expressed in <b>all</b> the selected conditions.
                    </p>
                </div>
            </div>

        </div>
</div>
</div>


<a href="#0" class="cd-top" title="Back to Top">Top</a>
{% endblock content %}



{% block scripts %}
    <!-- CHANGE ACTIVE NAVBAR BUTTON -->
    <script>
        $(".navbar-element").toggleClass('active inactive');
        $("#navbar-tutorial").toggleClass('inactive active')
    </script>

{% endblock scripts%}
