{% extends 'NetExplorer/base.html' %}
{% load static %}

{% block title %}- Tutorial{% endblock title %}



{% block content %}

<div class="row">
    <nav class="col-sm-2" id="myScrollspy">
        <ul class="nav nav-pills nav-stacked" data-spy="affix" data-offset-top="205">
            <li><a href="#genesearch">Gene Search</a></li>
            <li><a href="#blast">BLAST</a></li>
            <li><a href="#netexplorer">NetExplorer</a></li>
            <li><a href="#pathwayfinder">Pathway Finder</a></li>
            <hr>
            <li id="back_home"><a href="/" ><span class="glyphicon glyphicon-triangle-left"></span>Go back Home</a></li>
        </ul>
    </nav>

    <div class="col-sm-10">

      <!-- GENE SEARCH TUTORIAL -->
      <div class="tutorial-panel panel panel-info">
        <div class="panel-heading"><a class="tutorial-title" name="genesearch"><h3>Gene Search</h3></a></div>
        <div class="panel-body">
            Lorem ipsum dolor sit amet, consectetur adipiscing elit. Morbi velit ante, volutpat eu tincidunt at, laoreet ut arcu. Fusce porttitor sem dui, sed facilisis augue tempus nec. Proin pharetra porta neque, vitae venenatis libero viverra at. Suspendisse mattis, est vitae dignissim volutpat, magna magna pulvinar nisi, non cursus tortor augue vestibulum tellus. Quisque fringilla augue at euismod laoreet. Vestibulum viverra sed dolor id tempus. Nunc eleifend ligula eu lacus rhoncus sodales. Fusce id tellus aliquam, vulputate ex ac, consectetur est. Nunc feugiat, urna nec ornare fringilla, urna turpis rhoncus eros, id efficitur neque tellus id augue. Vivamus id odio eu justo gravida ornare.
        </div>
      </div>

       <!-- BLAST TUTORIAL -->
      <div class="tutorial-panel panel panel-blast">
        <div class="panel-heading"><a class="tutorial-title" name="blast"><h3>BLAST</h3></a></div>
        <div class="panel-body">
            Lorem ipsum dolor sit amet, consectetur adipiscing elit. Morbi velit ante, volutpat eu tincidunt at, laoreet ut arcu. Fusce porttitor sem dui, sed facilisis augue tempus nec. Proin pharetra porta neque, vitae venenatis libero viverra at. Suspendisse mattis, est vitae dignissim volutpat, magna magna pulvinar nisi, non cursus tortor augue vestibulum tellus. Quisque fringilla augue at euismod laoreet. Vestibulum viverra sed dolor id tempus. Nunc eleifend ligula eu lacus rhoncus sodales. Fusce id tellus aliquam, vulputate ex ac, consectetur est. Nunc feugiat, urna nec ornare fringilla, urna turpis rhoncus eros, id efficitur neque tellus id augue. Vivamus id odio eu justo gravida ornare.
        </div>
      </div>

       <!-- NETEXPLORER TUTORIAL -->
      <div class="tutorial-panel panel panel-success">
        <div class="panel-heading"><a class="tutorial-title" name="netexplorer"><h3>Net Explorer</h3></a></div>
        <div class="panel-body">
            <p>
                NetExplorer will allow you to explore the predicted networks dinamically, adding and removing nodes from several interactomes.
                Use the menus on the right to manipulate the graph, which will display on the left side of the screen. Each node on the graph
                will represent a contig (for planarian datasets) or a protein (for the Human reference dataset). These nodes will be connected by edges (lines),
                and each type of them will have a different meaning:
            </p>
            <img class="tutorial-img" src="{% static 'Images/tutorial/legend.png' %}"/>

            <br>
            <h4>Search functionality</h4>
            <img class="tutorial-subtitle-netexplorer" src="{% static 'Images/tutorial/underline-netexplorer.png' %}"/>
            <br>
            <img class="tutorial-img" src="{% static 'Images/tutorial/search.png' %}"/>
            <p>
                In order to search for contigs, you have to select a Dataset (a planarian transcriptome for which we predicted an interactome), and then look for contigs.
                You can use 4 types of symbols to do that:
            </p>
            <ul>
                <li class="tutorial-li">
                    <strong>Contig identifiers</strong> (which will be specific to each transcriptome). E.g. <i>cth1_Trc_comp11081_c0_seq1</i> (for dataset <i>Cthulhu</i>)
                </li>

                <li class="tutorial-li">
                    <strong>Human protein symbols</strong>. This will retrieve all the homologous contigs to the specified human protein. E.g. <i>THOC2</i>.
                    You can use the wildcard <i>*</i>. E.g. <i>THOC*</i> will display the nodes homologous to <i>THOC, THOC1, THOC2, THOC3, THOC4, THOCXX, etc.</i>
                </li>

                <li class="tutorial-li">
                    <strong>Gene Ontology identifiers</strong>. This will display all the planarian contigs with a human homolog with the specified GO code.
                    E.g. <i>GO:0000002</i>.
                </li>

                <li class="tutorial-li">
                    <strong>PFAM identifiers</strong>. This will retrieve all the planarian contigs with the specified PFAM annotated through our protocol (that is, with HMMER with a e-value cutoff of 10<sup>-10</sup>).
                    E.g. <i>PF01369.15</i>.
                </li>
            </ul>

            <p>
                You can do multiple searches at once, using commas ',' to separate each item. You can mix the different types of symbols in the same search, e.g. <i>PF01369.15,GO:0000002</i>.
            </p>
            <p>
                Once you have typed the symbols you want to look for, and have selected the appropiate dataset, click on the button <kbd class="tutorial-highlight-netexplorer">+ Add</kbd>.
                Note that you will not lose the network already displayed on the left, so you can add nodes individually without losing the visualization.
            </p>
            <p>
                All the nodes for which you have searched (either using this search function or by "expanding" them, see below) will be outlined in green,
                so that you can remember them.
            </p>

            <br>
            <h4>Basic controls</h4>
            <img class="tutorial-subtitle-netexplorer" src="{% static 'Images/tutorial/underline-netexplorer.png' %}"/>
            <br>

            <h4>1. Layout</h4>
            <img class="tutorial-img" src="{% static 'Images/tutorial/layout.png' %}"/>
            <p>Using this dropdown menu you can select the layout of the network. When you select it, the layout will be run on the display,
                and the positions of the nodes will change. Take note that some layouts will change each time you run them. There are 6 available layouts right now:</p>
            <img class="tutorial-img" src="{% static 'Images/tutorial/layouts.png' %}"/>
            <br>
            <hr>
            <h4>2. Show homologs</h4>
            <img class="tutorial-img" src="{% static 'Images/tutorial/homologs.png' %}"/>
            <p>
                When changing this switch to <kbd class="tutorial-highlight-netexplorer">On</kbd>, the homologs of the displayed planarian contigs will be drawn on the visualization.
                They will appear as blue diamonds. They behave as all the other nodes, but with the exception that you can't <kbd class="tutorial-highlight-netexplorer">Expand</kbd> them.
                Turning this option <kbd class="tutorial-highlight-netexplorer">On</kbd> or <kbd class="tutorial-highlight-netexplorer">Off</kbd> will <strong>not</strong> change the layout of the planarian nodes.
            </p>
            <img class="tutorial-img" src="{% static 'Images/tutorial/homolog.png' %}"/>
            <br>
            <hr>
            <h4>3. Connections manipulation</h4>
            <img class="tutorial-img" src="{% static 'Images/tutorial/connections.png' %}"/>
                 <br>
                <kbd class="tutorial-highlight-button">Show connections</kbd>: <br>
                    <p>
                        This button will add all the interactions between the nodes displayed on the visualization. It will NOT add any node.
                    </p>
                <br>
                <kbd class="tutorial-highlight-button">Remove Unconnected</kbd>: <br>
                    <p>
                        Activating this button will remove all the nodes without a <strong>visible</strong> connection.
                        Note that this applies to all types of connection (homology or interactions).
                        Use together with <kbd class="tutorial-highlight-button">Filter interactions</kbd> to filter networks.
                    </p>
            <br>

            <hr>
            <h4>4. Filter interactions</h4>
            <img class="tutorial-img" src="{% static 'Images/tutorial/filter.png' %}"/>
            <p>
                With this slider you can hide the interactions using the percentage of votes of the random forest classifier as a cut-off.
                Note that the hidden interactions are not removed, so you can get them back by using another cut-off.
            </p>
            <p>
                Right now, the only available cut-offs are: <i>0.6, 0.65, 0.70, 0.75, 0.8, 0.85, 0.9</i> and <i>0.95</i>.
                You can use this function in conjuction with the <kbd class="tutorial-highlight-button">Remove Unconnected</kbd> button to filter the graphs; first filtering the
                interactions by the desired cut-off, and then removing all the unconnected nodes.
                This is useful when wanting to work with a more accurate network, or when the nodes displayed on the visualization is too high.
            </p>
            <br>


            <h4>Click behaviour</h4>
            <img class="tutorial-subtitle-netexplorer" src="{% static 'Images/tutorial/underline-netexplorer.png' %}"/>
            <br>

            <img class="tutorial-img" src="{% static 'Images/tutorial/behaviour.png' %}"/>
            <p>
                If you have nodes on the visualization, you can determine what the application will do when you click on them.
                There are 3 possible behaviours, and you can change them by clicking on the appropiate one:
            </p>
            <hr>
            <h4>1. Card</h4>
            <p>
                When this behaviour is set, an information card for the clicked contig/protein will appear on the screen.
                These cards are the same displayed in the <a href="#genesearch">Gene Search</a> application.
            </p>
            <hr>
            <br>
            <h4>2. Expand</h4>
            <img class="tutorial-img" src="{% static 'Images/tutorial/expand.png' %}"/>
            <p>
                When clicking on nodes you will add all the interactions for such node to the visualization.
            </p>
            <hr>
            <br>
            <h4>3. Delete</h4>
            <p>
                Delete the clicked node and all its interactions. Note that all the other nodes displayed on the graph
                will NOT be removed, even if they have no other interactions.
            </p>

            <hr>
            <br>
            <h4>Expression Data</h4>
            <img class="tutorial-subtitle-netexplorer" src="{% static 'Images/tutorial/underline-netexplorer.png' %}"/>
            <br>
            <img class="tutorial-img" src="{% static 'Images/tutorial/mapexpression.png' %}"/>
            <p>
                If you click on the <kbd class="tutorial-highlight-button">Map Expression</kbd> button, the dialog showed on the above image will appear.
                Using this dialgo, you will be able to select a Expression experiment and map the expression values over the displayed nodes on the visualization.
                There are two ways in which you can do this: mapping the values of one sample as a color gradient, or mapping the logFC of two samples as a divergent color gradient.
                Make sure you are selecting two different samples in the <kbd class="tutorial-highlight-button">Two Samples</kbd> mode.
            </p>
            <img class="tutorial-img" src="{% static 'Images/tutorial/samples.png' %}"/>
            <p>
                Finally, select a color gradient from the available presets and click <kbd class="tutorial-highlight-netexplorer">OK</kbd>.
                The process of fetching the expression values and coloring the nodes may take a while depending on the number of nodes displayed.
                None of this will change the layout of the nodes, so feel free to experiment with the colors and the samples.
                When the process is done, a legend will appear on the visualization:
            </p>
            <img class="tutorial-img" src="{% static 'Images/tutorial/expressionlegend.png' %}"/>
            <p>
                If you are in <kbd class="tutorial-highlight-button">One Sample</kbd> mode, each color will represent the percentil
                to which the expression value belongs; for example, the color for the percentil 10% indicates that the nodes colored like that
                have a expression value lower than 95% to 90% of the genes in the gene expression experiment.
            </p>

            <p>
                If you are in <kbd class="tutorial-highlight-button">Two Samples</kbd> mode, each color will represent the log2FoldChange between
                the selected <i>sample 1</i> and <i>sample 2</i>.
            </p>

            <p>
                In either case, the nodes colored as NA are contigs for which there is no expression information for that particular experiment and sample.
            </p>
            <p>
                When a expression experiment has been mapped onto the visualization you can keep ading nodes normally. However, if you want the expression
                to be mapped to the new added nodes, you have to open the dialog again and re-run the process. Don't expect new nodes to be colored according to
                the expression information previously mapped.
            </p>
            <br>


            <h4>General Controls</h4>
            <img class="tutorial-subtitle-netexplorer" src="{% static 'Images/tutorial/underline-netexplorer.png' %}"/>
            <br>
            <img class="tutorial-img" src="{% static 'Images/tutorial/controls.png' %}"/>
            <p>
                These simple buttons will allow the user to save, import and export the visualization in several formats, as well as centering it to the screen.
            </p>

            <kbd class="tutorial-highlight-button">
                <span class="glyphicon glyphicon-resize-small" aria-hidden="true"></span>
            </kbd>&nbsp;&nbsp; Center the visualization when clicked.
            <br><br>

            <kbd class="tutorial-highlight-button">
                <span class="glyphicon glyphicon-picture" aria-hidden="true"></span>
            </kbd>&nbsp;&nbsp; Saves a png image of the visualization. Please, take note that the expression legend will NOT be saved, as it is not part of the visualization.
            <br><br>

            <kbd class="tutorial-highlight-button">
                <span class="glyphicon glyphicon-share" aria-hidden="true"></span>
            </kbd>&nbsp;&nbsp; Exports the graph (nodes and edges) as a tabular file. This graph can then be imported to Cytoscape, or Gephi, or other programs to perform
            further analyses.
            <br><br>

            <kbd class="tutorial-highlight-button">
                <span class="glyphicon glyphicon-save" aria-hidden="true"></span>
            </kbd>&nbsp;&nbsp; Saves the graph as a JSON file. This is the format used by NetExplorer. If you want to save a particular visualization to work on it later in
            NetExplorer, use this button. All the positions, colors, nodes and edges will be saved and they can be restored.
            <br><br>

            <kbd class="tutorial-highlight-button">
                <span class="glyphicon glyphicon-open" aria-hidden="true"></span>
            </kbd>&nbsp;&nbsp; Imports a JSON file (exported by using the previous button). All the visualization will be restored. Any network on the visualization previous
            to importing the JSON file will be lost, so please be careful.
            <br><br>

            <kbd class="tutorial-highlight-button">
                <span class="glyphicon glyphicon-trash" aria-hidden="true"></span>
            </kbd>&nbsp;&nbsp; Remove everything (nodes, edges, colors, legends) from the visualization. This process is <strong>irreversible</strong>.
            <br><br>



        </div>
      </div>

       <!-- PATHWAY FINDER TUTORIAL -->
      <div class="tutorial-panel panel panel-warning">
        <div class="panel-heading"><a class="tutorial-title"  name="pathwayfinder"><h3>Pathway Finder</h3></a></div>
        <div class="panel-body">
            Lorem ipsum dolor sit amet, consectetur adipiscing elit. Morbi velit ante, volutpat eu tincidunt at, laoreet ut arcu. Fusce porttitor sem dui, sed facilisis augue tempus nec. Proin pharetra porta neque, vitae venenatis libero viverra at. Suspendisse mattis, est vitae dignissim volutpat, magna magna pulvinar nisi, non cursus tortor augue vestibulum tellus. Quisque fringilla augue at euismod laoreet. Vestibulum viverra sed dolor id tempus. Nunc eleifend ligula eu lacus rhoncus sodales. Fusce id tellus aliquam, vulputate ex ac, consectetur est. Nunc feugiat, urna nec ornare fringilla, urna turpis rhoncus eros, id efficitur neque tellus id augue. Vivamus id odio eu justo gravida ornare.
        </div>
      </div>


    </div>
</div>

{% endblock content %}



{% block scripts %}
    <!-- CHANGE ACTIVE NAVBAR BUTTON -->
    <script>
        $(".navbar-element").toggleClass('active inactive');
        $("#navbar-tutorial").toggleClass('inactive active')
    </script>

{% endblock scripts%}
